samples/core/callback/main.cpp

```diff
@@ -0,0 +1,341 @@
+            };
+            const std::string filename =
+                "callbackcpp_out" + std::to_string(job_id) + ".png";
+            cl::sdk::write_image(filename.c_str(), image);
+        });
+
+    // job.buffer goes out of scope here, releasing the OpenCL buffer.
+}
+
+struct CallbackOptions
+{
+    std::size_t side;
+    std::size_t iterations;
+    std::size_t write_iter;
+};
+
+} // namespace
+
+template <> auto cl::sdk::parse<CallbackOptions>()
+{
+    return std::make_tuple(
+        std::make_shared<TCLAP::ValueArg<std::size_t>>(
+            "s", "side", "Side length of the generated image in pixels", false,
+            512, "positive integral"),
+        std::make_shared<TCLAP::ValueArg<std::size_t>>(
+            "i", "iter", "Number of iterations in the simulation", false, 10000,
+            "positive integral"),
+        std::make_shared<TCLAP::ValueArg<std::size_t>>(
+            "w", "write_iter",
+            "Controls after how many iterations the intermediate result is "
+            "written to file",
+            false, 1000, "positive integral"));
+}
+
+template <>
+CallbackOptions cl::sdk::comprehend<CallbackOptions>(
+    std::shared_ptr<TCLAP::ValueArg<size_t>> side_arg,
+    std::shared_ptr<TCLAP::ValueArg<size_t>> iter_arg,
+    std::shared_ptr<TCLAP::ValueArg<size_t>> write_iter_arg)
+{
+    return CallbackOptions{ side_arg->getValue(), iter_arg->getValue(),
+                            write_iter_arg->getValue() };
+}
+
+int main(int argc, char* argv[])
+{
+    try
+    {
+        // Parse command-line options
+        auto opts =
+            cl::sdk::parse_cli<cl::sdk::options::Diagnostic,
+                               cl::sdk::options::SingleDevice, CallbackOptions>(
+                argc, argv);
+        const auto& diag_opts = std::get<0>(opts);
+        const auto& dev_opts = std::get<1>(opts);
+        const auto& alg_opts = std::get<2>(opts);
+
+        // Create runtime objects based on user preference or default
+        cl::Context context = cl::sdk::get_context(dev_opts.triplet);
+        cl::Device device = context.getInfo<CL_CONTEXT_DEVICES>().at(0);
+
+        // There is a separate command queue for kernel launches,
+        // device->device copies and device->host copies.
+        // Synchronization is ensured via events (see later).
+        cl::CommandQueue compute_queue{ context, device };
+        cl::CommandQueue copy_queue{ context, device };
+        cl::CommandQueue read_queue{ context, device };
+
+        cl::Platform platform{
+            device.getInfo<CL_DEVICE_PLATFORM>()
+        }; // https://github.com/KhronosGroup/OpenCL-CLHPP/issues/150
+
+        if (!diag_opts.quiet)
+        {
+            std::cout << "Selected platform: "
+                      << platform.getInfo<CL_PLATFORM_VENDOR>() << "\n"
+                      << "Selected device: " << device.getInfo<CL_DEVICE_NAME>()
+                      << "\n"
+                      << std::endl;
```

Comment:
I don't care too strongly about this, but if we want the C++ version to output the same information as the C version we should also print the OpenCL C version here.  See: `cl_util_print_device_info`

We could also just file an issue for this and (potentially) fix it later.

Reply:
Done as requested.